src/gridcoin/gridcoin.cpp

```diff
@@ -195,6 +195,60 @@ void InitializeExplorerFeatures()
     fExplorer = GetBoolArg("-scraper", false) && GetBoolArg("-explorer", false);
 }
 
+//!
+//! \brief Check the block index for unusual entries.
+//!
+//! This runs a simple check on the entries in the block index to determine
+//! whether the index contains invalid state caused by an unclean shutdown.
+//! This condition was originally detected by an assertion in a routine for
+//! stake modifier checksum verification. Because Gridcoin removed modifier
+//! checksums and checkpoints, we reinstate that assertion here as a formal
+//! inspection.
+//!
+//! This function checks that no block index entries contain a null pointer
+//! to a previous block. The symptom may indicate a deeper problem that can
+//! be resolved by tuning disk synchronization in LevelDB. Until then, this
+//! heuristic has proven itself to be effective for identifying a corrupted
+//! database.
+//!
+void CheckBlockIndexJob()
+{
+    LogPrintf("Gridcoin: checking block index...");
+
+    bool corrupted = false;
+
+    if (pindexGenesisBlock) {
+        LOCK(cs_main);
+
+        for (const auto& index_pair : mapBlockIndex) {
+            const CBlockIndex* const pindex = index_pair.second;
+
+            if (!pindex || !(pindex->pprev || pindex == pindexGenesisBlock)) {
+                corrupted = true;
+                break;
+            }
+        }
```

Comment:
We are putting this on the scheduler thread, which is initialized late IIRC. Is there any problem with holding the lock on main for a long time given the timing of the job?

Reply:
None that I've come across. At least for now, the job executes very quickly on modern hardware (about a second). The lock may hold up net messages for blocks and transactions briefly while the job runs, but I don't think it will cause any problems.

Reply:
Sounds good to me...